### PR TITLE
Update to Geth v1.12.2 Mawinor

### PR DESCRIPTION
**Pull request by protolambda**
**Description**

Update to v1.12.2

Notable changes with v1.12.2:
- Tx pool was split into two sub-pools.
- Verkle-override flag was added, causing common merge conflicts with optimism override flags.
- RPC namespaces have been split into separate files.
- Optimism diff: preserve alias "pending" to "latest": upstream returns an error when the state is not available.
- `backend.Attach()` does not return any error anymore, instead of always-nil error.
- `v5disc` flag deprecated in favor of `discv5`
- `func (pool *LegacyPool) Filter(tx *types.Transaction) bool` now prevents the deposit transactions of a block from being inserted back into the pool after reorg, reducing our diff a tiny bit.
- `pool.l1CostFn` now does not take the isDepositTx bool anymore, as the pool already filters those out anyway.

Minor fixes:
- genesis loading updated to preserve stateHash
- ethconfig config TOML format encoding/decoding updated to include optimism rollup attributes

Includes updated `fork.yaml`. The op-geth diff site will auto-build with CI when merged. To locally build, simply run `forkdiff` in the root of the repository. (install: `go install github.com/protolambda/forkdiff@latest`)

### CODE REVIEW

File: core/txpool/legacypool/legacypool.go [hunk 1]
```diff
@@ -0,0 +1,1996 @@
...
+// executable/pending queue and any subsequent transactions that become unexecutable
+// are moved back into the future queue.
+//
+// Note: transactions are not marked as removed in the priced list because re-heaping
+// is always explicitly triggered by SetBaseFee and it would be unnecessary and wasteful
+// to trigger a re-heap is this function
+func (pool *LegacyPool) demoteUnexecutables() {
+	// Iterate over all accounts and demote any non-executable transactions
+	gasLimit := pool.currentHead.Load().GasLimit
+	// TODO
```
Comment (Inphi) [inline core/txpool/legacypool/legacypool.go:1677]: TODO?

File: core/txpool/legacypool/legacypool.go [hunk 2]
```diff
@@ -0,0 +1,1996 @@
...
+		balance := pool.currentState.GetBalance(addr)
+		if !list.Empty() && pool.l1CostFn != nil {
+			// Reduce the cost-cap by L1 rollup cost of the first tx if necessary. Other txs will get filtered out afterwards.
+			el := list.txs.FirstElement()
+			if l1Cost := pool.l1CostFn(el.RollupDataGas()); l1Cost != nil {
+				balance = new(big.Int).Sub(balance, l1Cost) // negative big int is fine
+			}
+		}
+		// Drop all transactions that are too costly (low balance or out of gas), and queue any invalids back for later
+		drops, invalids := list.Filter(balance, gasLimit)
```
Comment (Inphi) [inline core/txpool/legacypool/legacypool.go:1697]: This should be a bit more restrictive and consider the effective gas limit.

Comment (Inphi) [inline core/txpool/legacypool/legacypool.go:1697]: This should be a bit more restrictive and consider the effective gas limit.

File: eth/backend.go [hunk 3]
```diff
@@ -229,11 +233,20 @@ func New(stack *node.Node, config *ethconfig.Config) (*Ethereum, error) {
...
+	}
+	blobPool := blobpool.New(config.BlobPool, eth.blockchain)
+
 	if config.TxPool.Journal != "" {
 		config.TxPool.Journal = stack.ResolvePath(config.TxPool.Journal)
 	}
-	eth.txPool = txpool.NewTxPool(config.TxPool, eth.blockchain.Config(), eth.blockchain)
+	legacyPool := legacypool.New(config.TxPool, eth.blockchain)
 
+	eth.txPool, err = txpool.New(new(big.Int).SetUint64(config.TxPool.PriceLimit), eth.blockchain, []txpool.SubPool{legacyPool, blobPool})
```
Comment (Inphi) [inline eth/backend.go:246]: We shouldn't allow blob txs in the pool. Even if ingress filters them.

**Review by Inphi — COMMENTED** [review]
Looks mostly good. Just a couple comments regarding the txpool.

File: eth/backend.go (hunk 3 above)
Comment (protolambda) [reply eth/backend.go:246]: Hmm, I assumed that blobs won't be accepted before the dencun fork is active? Is that not the case? I can add an explicit filter for the op-stack chains to not accept blobs into the tx pool.

Comment (Inphi) [reply eth/backend.go:246]: you're right. It's asserted by the blobpool while validating transactions. This is fine.

File: core/txpool/legacypool/legacypool.go (hunk 2 above)
Comment (protolambda) [reply core/txpool/legacypool/legacypool.go:1697]: The gasLimit is adjusted at the start of `demoteUnexecutables` (the one with the TODO, fixed now) / `promoteExecutables`

Comment (protolambda) [reply core/txpool/legacypool/legacypool.go:1697]: same here

File: miner/payload_building.go [hunk 4]
```diff
@@ -163,23 +163,35 @@ func (payload *Payload) ResolveFull() *engine.ExecutionPayloadEnvelope {
...
+	return engine.BlockToExecutableData(payload.full, payload.fullFees, nil, nil, nil)
 }
 
 // buildPayload builds the payload according to the provided parameters.
 func (w *worker) buildPayload(args *BuildPayloadArgs) (*Payload, error) {
 	// Build the initial version with no transaction included. It should be fast
 	// enough to run. The empty payload can at least make sure there is something
 	// to deliver for not missing slot.
-	empty, _, err := w.getSealingBlock(args.Parent, args.Timestamp, args.FeeRecipient, args.Random, args.Withdrawals, true, args.Transactions, args.GasLimit)
+	empty, emptyFees, err := w.getSealingBlock(args.Parent, args.Timestamp, args.FeeRecipient, args.Random, args.Withdrawals, true, args.Transactions, args.GasLimit)
```
[the remark this answers is not in the document]
Comment (protolambda) [reply miner/payload_building.go:185]: I'll add a comment to the code about the "empty" term referring to no tx-pool usage, matching upstream intents

**Review by Inphi — APPROVED** [review]
LGTM! I'll take a look at the other one later today.

**Pull request merged**